pkg/build/builder/dockerutil.go

```diff
@@ -49,12 +60,23 @@ func retryImageAction(actionName string, action func() error) error {
 		time.Sleep(DefaultPushOrPullRetryDelay)
 	}
 
-	switch errors.Cause(err) {
-	case idocker.ErrUnauthorizedForCredentials:
-		// strip off wrappers that mainly add the image name as their added context
-		err = errors.Cause(err)
+	if errs, ok := errors.Cause(err).(errcode.Errors); ok {
+		// if this error is a group of errors, process them all in turn
+		var unwrap bool
+		for i := range errs {
+			if err, ok := errs[i].(errcode.Error); ok {
+				if err.Code == errcode.ErrorCodeUnauthorized {
+					unwrap = true
+				}
+			}
+		}
+		if unwrap {
+			err = errors.Cause(err)
```

Comment:
isn't this stomping the original group of errors?  so now you'll only log a single (unwrapped)error from the original group?

Reply:
The scoping here is probably confusing - the `errors.Cause()` call on line 74 is removing additional context (that `errors.Wrapf()` would have added) from the `err` defined on line 52, when we detect that there's an authorization error in there somewhere.  If the result that first went into `errors.Wrapf()` is ultimately an `errcode.Errors`, that list is what `err` will be set to on line 74.  I'll change the name of the one we create on line 67.

Reply:
sorry i'm still a little confused:
`if errs, ok := errors.Cause(err).(errcode.Errors); ok`

if that's true, then `err = errors.Cause(err)` is going to set err== something that is an array of errors (errcode.Errors)

then you're going to pass array of errors (interface errcode.Errors) down to unwrapUnauthorizedError

which then calls `cause := errors.Cause(err)`.  What does errors.Cause(err) return when err is an errcode.Errors?

I would have expected logic that iterates the array and unwraps each err within the array that needs unwrapping.

Reply:
`errors.Cause()` looks at the error that `err` wraps, repeating until it reaches an error that isn't a wrapper (i.e., an object that doesn't provide a `Cause()` method), and `errcode.Errors` does not provide such a method, so if it's an `errcode.Errors`, and one of the errors in its list is an "unauthorized" error that we recognize, we're removing the layers of wrapping around the errors list.

Reply:
ok.....so if action() returned an errors list or a wrapped errors list, you will scan the errors list, see if it contains an "unauthorized error", and then attempt to unwrap the list (leaving the list itself).

if that's the intent (as compared w/ unwrapping the unauthorized error specifically), then this lgtm